### PR TITLE
Add function composition

### PR DESCRIPTION
**Pull request by fkleedorfer**

### CODE REVIEW

**Comment by fkleedorfer** [comment]
Having experimented with the fnoc ontology further, I am really on the fence whether `fnoc:CompositionMappingSource` and `fnoc:CompositionMappingTarget` should be the same class - which I would call `fnoc:CompositionMappingEndpoint`. The classes are implicit and do not appear in any written mappings, so you'll only see them if you do rdfs reasoning or SHACL. 

Argument for two distinct classes:
* if you do reasoning/SHACL you can differentiate source and target by their class. Not sure what you'd need that for, though. Moreover, if an 'endpoint' is used as target and source by two different mappings, it has both classes. Is that what we want?
 
Arguments for one class:
* Makes things simpler. 

Leaning toward one class.

File: section/abstract.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ The preferred prefix is `fno:`
...
 The discussed ontology's version is [1.0.0](https://w3id.org/function/ontology/1.0.0).
 
-The used vocabularies are the
-[FnO mapping vocabulary v0.2.1](https://w3id.org/function/vocabulary/mapping/0.2.1) (`fnom:`)
-and the
-[FnO implementation vocabulary v0.1.1](https://w3id.org/function/vocabulary/implementation/0.1.1) (`fnoi:`).
+The used vocabularies are 
+* [FnO mapping vocabulary v0.2.1](https://w3id.org/function/vocabulary/mapping/0.2.1) (`fnom:`)
+* [FnO implementation vocabulary v0.1.1](https://w3id.org/function/vocabulary/implementation/0.1.1) (`fnoi:`).
+* [FnO composition vocabulary v0.1.1](https://w3id.org/function/vocabulary/composition/0.1.1) (`fnoc:`).
```
Comment (bjdmeest) [inline section/abstract.md]: I guess this should be 0.1.0 :)

File: section/definitions.md [hunk 2]
```diff
@@ -51,6 +51,11 @@ Furthermore, we define following concepts:
   _The ontology term is defined at [fno:Implementation](#fno-implementation)._
 - A <dfn>mapping</dfn> links (abstract) <a data-lt="function">functions</a> and (concrete) <a data-lt="implementation">implementations</a>.
   _The ontology term is defined at [fno:Mapping](#fno-Mapping)._
+- A <dfn>composition</dfn> combines multiple functions by mapping <a data-lt="parameter">parameters</a> or <a data-lt="output">outputs</a>  
+of <a data-lt="function">constituent functions</a> onto one another, defining a nesting or concatenation of functions.
+  _The ontology term is defined at [fnoc:Composition](#fnoc-Composition). The composition consists of a set of CompositionMappings.
```
Comment (bjdmeest) [inline section/definitions.md]: I think a `_` is missing after [fnoc:Composition]

File: section/ontology.md [hunk 3]
```diff
@@ -413,3 +413,238 @@ ex:leftPadMapping
     fno:returnMapping    [ a                   fnom:DefaultReturnMapping ;
                            fnom:functionOutput ex:outputStringOutput ] .
 ```
+
+### Function Composition {#composition}
+
+
```
Comment (bjdmeest) [inline section/ontology.md:420]: I'd add a `> Function Composition is experimental and subject to change` disclaimer here

File: section/ontology.md [hunk 4]
```diff
@@ -413,3 +413,238 @@ ex:leftPadMapping
...
+We will realize `ex:sum3Function` (defined in the previous paragraph) using `ex:sumFunction` introduced earlier, as illustrated with the following pseudo code: 
+```
+    ex:sum3Function(intParameterA, intParameterB, intParameterC) =  
+        ex:sumFunction(ex:sumFunction(intParameterA, intParameterB), intParameterC) 
+```
+
+In this construct, we use `ex:sumFunction` twice. In order to be able to reference the two distinct calls of the
+function `ex:sumFunction` unambiguously, we first create two distinct *applications* 
+of it using <a href="#fnoc-applies">`fnoc:applies`</a>. Then, we define the composition using a set of <a>composition mapping</a>s.
```
Comment (bjdmeest) [inline section/ontology.md]: Is this application always needed? Or only in the case of using the same function multiple times?

File: section/ontology.md [hunk 5]
```diff
@@ -413,3 +413,238 @@ ex:leftPadMapping
...
+### Partial Function Application {#partialFunctionApplication}
+
+In the following example, we define the function `ex:add10` as a partial application of `ex:sumFunction` by providing 
+the constant value `10` for the parameter `ex:intParameterA`. The parameter value is specified using the associated 
+<a href="#fn-predicate">fno:predicate</a> `ex:startValue`. 
+
+```turtle "example": " " 
+ex:add10 a fnoc:PartiallyAppliedFunction;
+    fnoc:partiallyApplies :sumFunction ;
+    ex:startValue 10 .
```
Comment (bjdmeest) [inline section/ontology.md]: Hmmm, I'm not really a fan of directly applying the predicate to the function, that's kind of 'crossing boundaries'. Would something like below be an option?

```turtle
ex:add10 a fnoc:PartiallyAppliedFunction;
    fnoc:partiallyApplies :sumFunction ;
    fnoc:partialApplication [
      fnoc:mapFromTerm 10 ;
      fnoc:mapToParameter ex:inParameterA
    ] .
```

**Comment by bjdmeest** [comment]
Given there are 2 different predicates, you'll always be able to differentiate between source and target, so I don't see a problem with using one class. I'm in favor of defining only what is needed, and extend later if necessary.  (considering your point about 2 classes: I don't see that as a problem :) ).

The PR looks good! I'd like to experiment a bit with some code to see whether there are obvious gaps we're missing before merging, if that's fine for you. Otherwise, we could label it as 'experimental' and already merge :)

Meanwhile, I found https://commons.wikimedia.org/wiki/File:Boolean_Algebra_with_very_early_Wikilambda.webm
There, it seems an implementation and a composition are put on a sibling level, which feels similar to what we came up with, so that gives me confidence we're on the right track :). Their structure is more hierarchical, but ours (being flat) is - I think - more flexible, certainly given we take multiple returns into account, so that difference is fine imo

File: section/ontology.md (hunk 4 above)
Comment (fkleedorfer) [reply section/ontology.md]: I think it is only needed if we want to use the same function multiple times in a composition. Will clarify in the text.

File: section/ontology.md (hunk 5 above)
Comment (fkleedorfer) [reply section/ontology.md]: Yes, that's good. Still concise compared to the full composition equivalent, and ontologically clean.

File: section/abstract.md (hunk 1 above)
Comment (fkleedorfer) [reply section/abstract.md]: done

File: section/definitions.md (hunk 2 above)
Comment (fkleedorfer) [reply section/definitions.md]: done

File: section/ontology.md (hunk 3 above)
Comment (fkleedorfer) [reply section/ontology.md:420]: done - I think. Maybe check if what I did is what you wanted.

File: section/ontology.md (hunk 4 above)
Comment (fkleedorfer) [reply section/ontology.md]: done

File: section/ontology.md (hunk 5 above)
Comment (fkleedorfer) [reply section/ontology.md]: On second thought, wouldn't it be easier to re-use the composedOf property?
```
ex:add10 a fnoc:PartiallyAppliedFunction;
    fnoc:partiallyApplies :sumFunction ;
    fnoc:composedOf [
      fnoc:mapFromTerm 10 ;
      fnoc:mapToParameter ex:inParameterA
    ] .
``` 
EDIT: scratch that. 

How about 
```
ex:add10 a fnoc:PartiallyAppliedFunction;
    fnoc:partiallyApplies :sumFunction ;
    fnoc:bindsParameter [
      fnoc:bindToTerm 10 ;
      fnoc:boundParameter ex:inParameterA
    ] .
```
However, this construction isn't as neat as the composition, yet. Might need more work.

Comment (bjdmeest) [reply section/ontology.md]: I'm fine with your latest suggestion. It might not be perfect yet, but it's at least something we can start from :)

**Comment by fkleedorfer** [comment]
> The PR looks good! I'd like to experiment a bit with some code to see whether there are obvious gaps we're missing before merging, if that's fine for you

sure, take your time!

**Comment by fkleedorfer** [comment]
Have you had a chance to take another look at it?

**Comment by bjdmeest** [comment]
Oh man, I have, I've actually been working on PoC-ing an FnO SDK to 'prove' the definitions work (see https://github.com/FnOio/function-handler-js/blob/master/src/FunctionHandler.test.ts for everything except the composition), but I haven't been able to complete the SDK for composition yet.

I'm going to stop being the bottleneck, I'll merge now and publish. The section states 'experimental', so we can refine as we go :). Thanks a huge lot for this!

**Comment by bjdmeest** [comment]
Online at https://fno.io/spec/#composition!

**Comment by fkleedorfer** [comment]
Oh, great! Thanks!

**Pull request merged**